src/Factory.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/*
+ * This file is part of the jolicode/elastically library.
+ *
+ * (c) JoliCode <coucou@jolicode.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace JoliCode\Elastically;
+
+use JoliCode\Elastically\Serializer\ContextBuilderInterface;
+use JoliCode\Elastically\Serializer\StaticContextBuilder;
+use Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Normalizer\ArrayDenormalizer;
+use Symfony\Component\Serializer\Normalizer\DateTimeNormalizer;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
+use Symfony\Component\Serializer\Serializer;
+use Symfony\Component\Serializer\SerializerInterface;
+
+final class Factory
+{
+    /* Elastically config keys */
+    public const CONFIG_BULK_SIZE = 'elastically_bulk_size';
+    public const CONFIG_CLIENT = 'elastically_client';
+    public const CONFIG_DENORMALIZER = 'elastically_denormalizer';
+    public const CONFIG_INDEX_CLASS_MAPPING = 'elastically_index_class_mapping';
+    public const CONFIG_INDEX_NAME_MAPPER = 'elastically_index_name_mapper';
+    public const CONFIG_INDEX_PREFIX = 'elastically_index_prefix';
+    public const CONFIG_MAPPINGS_DIRECTORY = 'elastically_mappings_directory';
+    public const CONFIG_SERIALIZER = 'elastically_serializer';
+    public const CONFIG_SERIALIZER_CONTEXT_BUILDER = 'elastically_serializer_context_builder';
+    public const CONFIG_SERIALIZER_CONTEXT_PER_CLASS = 'elastically_serializer_context_per_class';
+
+    private array $config;
+
+    private Client $client;
+    private IndexNameMapper $indexNameMapper;
+    private IndexBuilder $indexBuilder;
+    private Indexer $indexer;
+    private ResultSetBuilder $resultSetBuilder;
+    private ContextBuilderInterface $contextBuilder;
+    private SerializerInterface $serializer;
+    private DenormalizerInterface $denormalizer;
+
+    public function __construct(array $config = [])
+    {
+        $this->config = $config;
+    }
+
+    public function buildClient(): Client
+    {
+        return $this->client ??= $this->config[self::CONFIG_CLIENT] ?? new Client(
+            $this->config,
+            null,
+            null,
+            $this->buildBuilder(),
+            $this->buildIndexNameMapper()
+        );
+    }
+
+    public function buildIndexNameMapper(): IndexNameMapper
+    {
+        return $this->indexNameMapper ??= $this->config[self::CONFIG_INDEX_NAME_MAPPER] ?? new IndexNameMapper($this->config[self::CONFIG_INDEX_PREFIX] ?? null, $this->config[self::CONFIG_INDEX_CLASS_MAPPING] ?? []);
```

Comment:
I know the null coalescing operator is nice and everything, but the readability of this method does not benefit from it IMO :grimacing: